jplag/src/main/java/de/jplag/clustering/algorithm/GaussianProcess.java

```diff
@@ -0,0 +1,210 @@
+            throw new IllegalArgumentException(MessageFormat.format("X has different dimensions at index {0} ({2}) and index {1} ({3})",
+                    brokenIndex - 1, brokenXIndex, X.get(brokenIndex - 1).getDimension(), X.get(brokenIndex).getDimension()));
+        });
+        if (noise <= 0) {
+            throw new IllegalArgumentException(MessageFormat.format("noise must be strictly positive, got {0}", noise));
+        }
+        if (lengthScale.length != X.get(0).getDimension()) {
+            throw new IllegalArgumentException(
+                    MessageFormat.format("lengthScale is of different dimension {0} than X values {1}", lengthScale.length, X.get(0).getDimension()));
+        }
+        double mean = 0;
+        double standardDeviation = 1;
+        if (normalize) {
+            mean = StatUtils.mean(Y);
+            standardDeviation = Math.sqrt(StatUtils.variance(Y, mean));
+            for (int i = 0; i < Y.length; i++) {
+                Y[i] = (Y[i] - mean) / standardDeviation;
+            }
+        }
+        RealVector yVector = new ArrayRealVector(Y);
+        RealVector scaleV = new ArrayRealVector(lengthScale);
+
+        RealMatrix k = maternKernel(X, scaleV);
+        k = k.add(new DiagonalMatrix(DoubleStream.generate(() -> noise).limit(X.size()).toArray(), false));
+
+        CholeskyDecomposition choDec = new CholeskyDecomposition(k);
+        RealVector w = choDec.getSolver().solve(yVector);
+
+        return new GaussianProcess(X, w, mean, standardDeviation, choDec, scaleV);
+    }
+
+    /**
+     * Matern kernel for nu=2.5 (we get a twice differentiable gp)
+     */
+    private static RealMatrix maternKernel(List<RealVector> X, RealVector lengthScale) {
+        RealMatrix k = new Array2DRowRealMatrix(X.size(), X.size());
+        for (int i = 0; i < X.size(); i++) {
+            RealVector left = X.get(i);
+            for (int j = i; j < X.size(); j++) {
+                RealVector right = X.get(j);
+                double maternVal = maternKernel(left, right, lengthScale);
+                k.setEntry(i, j, maternVal);
+                k.setEntry(j, i, maternVal);
+            }
+        }
+        return k;
+    }
+
+    private static RealVector maternKernel(List<RealVector> X, RealVector v, RealVector lengthScale) {
+        RealVector out = new ArrayRealVector(X.size());
+        for (int i = 0; i < X.size(); i++) {
+            out.setEntry(i, maternKernel(X.get(i), v, lengthScale));
+        }
+        return out;
+    }
+
+    private static double maternKernel(RealVector left, RealVector right, RealVector lengthScale) {
+        double dist = left.ebeDivide(lengthScale).getDistance(right.ebeDivide(lengthScale));
+        dist *= SQRT_5;
+        return (1 + dist) * Math.exp(-dist);
+    }
+
+    /**
+     * ascii graph for debugging
+     */
+    public String toString(RealVector min, RealVector max, int width, int height, double minY) {
+        char[][] out = new char[height][width];
```

Comment:
Maybe cache the string output since this method seems to be complex.

Reply:
The only use of this method is for creating debug output and then each output would be different.